# Script collection for the investigation of splice sites

These scripts provide functions to analyse splice sites based on a genome sequence (FASTA) and a corresponding annotation (GFF3) provided by the NCBI. Next, RNA-Seq reads can be mapped via STAR and the resulting BAM files can be converted into customized coverage files (COV) which are required for the validation of splice sites. The usage of individual splice sites can be quantified based on the number of mapped and propperly splitted RNA-Seq reads. These scripts generate report files and figures for the documentation of the analysis process.


# References

Boas Pucker and Samuel Fraser Brockington (2018). Genome-wide analyses supported by RNA-Seq reveal non-canonical splice sites in plant genomes. BMC Genomics. 2018;19(1). https://doi.org/10.1186/s12864-018-5360-z


Boas Pucker, Daniela Holtgr√§we, Bernd Weisshaar (2017). Consideration of non-canonical splice sites improves gene prediction on the Arabidopsis thaliana Niederzenz-1 genome sequence. BMC Research Notes, 10, 667. https://doi.org/10.1186/s13104-017-2985-y. 
